### ISSUE TITLE
STM: Support Scala.js

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#803]
Currently "synchronized" is used in the implementation of STM, which is not portable to Scala.js. This functionality should be factored out into a simple data structure:

```scala
class LockedRef[A] {
  def get: A
  def set(a: A): A 
  def lock: Unit
  def unlock: Unit
}
```

The Scala.js implementation will obviously not use locking.

The data structure can be placed in the jvm and js packages `scalaz.zio.internal`.

Then the use of `synchronized` inside `STM.scala` can be replaced by the above structure, bringing full portability of STM to Scala.js.

**Comment by heraklos** [comment]
I would like to take this issue :)

**Comment by mgibowski** [comment]
I just checked and found that the implementation also uses `java.util.concurrent.Semaphore` (`internal.globalLock`) and the spec uses `java.util.concurrent.CountDownLatch` - both of these are not implemented in Scala.js as they contain blocking operations. 

@jdegoes do you consider them portable in some way in the context of zio?
Or should they rather be extracted into other simple data structures?

**Comment by heraklos** [comment]
@jdegoes 
@mgibowski is right. When compiling Scala.js code containing STM method, the error `Referring to non-existent class java.uti.concurrent.Semaphore` was thrown. 
Am I supposed to create two data structure `LockedRef` and `Semaphore` to support STM in Scala.js and replace `synchronized` and `Semaphore` respectively?

**Comment by heraklos** [comment]
@jdegoes 
I have one more question.
When I just removed all `done synchronized`s from `atomically` in `STM.scala`, `STMSpec` still worked.
Does that mean we don't need `done synchronized` in the first place, or I should add some tests that fails when `atomically` is controlled exclusively?

**Comment by heraklos** [comment]
> When I just removed all done synchronizeds from atomically in STM.scala, STMSpec still worked.
Does that mean we don't need done synchronized in the first place, or I should add some tests that fails when atomically is controlled not exclusively?

It seems that removing `done synchronized` escapes test failures when there are not enough fibers forked to cause data confliction. 
I confirmed that changing the number of fibers from 10 to 100 caused test failure.
I may have to modify the number of fibers in some test case for STM.

**Comment by jdegoes** [comment]
@heraklos 

Thank you for taking this project on!

> I think it is possible to replace synchronized and Semaphore with just LockedRef data type because Semaphore being used in STM is binary semaphore and can be regarded as lock.

You are right, we only need `LockedRef`.

> When I just removed all done synchronizeds from atomically in STM.scala, STMSpec still worked.
Does that mean we don't need done synchronized in the first place, 

We do need it! The code will _not_ be correct without synchronization in the presence of multithreading. Of course, in Javascript, the code will always be correct since there is no multithreading. 😄

**Comment by heraklos** [comment]
@jdegoes 
Sorry for the slow work😭I got into a problem. 
Since I finally replaced synchronized and semaphore with LockedRef and I confirmed that my changes didn't break the existing test cases, I tried to make the code using STM work with scala.js.

But I got errors like the followings for each place HashMap.forEach is called in STM :
```
[error] Referring to non-existent method java.util.HashMap.forEach(java.util.function.BiConsumer)scala.Unit
[error]   called from zio.stm.STM$internal$.commitJournal(java.util.HashMap)scala.Unit
[error]   called from zio.stm.STM$internal$.tryCommit(zio.internal.Platform,scala.Function1)zio.stm.STM$internal$TryCommit
[error]   called from zio.stm.STM$.$$anonfun$atomically$1(scala.Function1,zio.internal.Platform)zio.ZIO
[error]   called from zio.stm.STM$.atomically(scala.Function1)zio.ZIO
[error]   called from zio.stm.STM$.commit$extension(scala.Function1)zio.ZIO
[error]   called from Main$.run(scala.collection.immutable.List)zio.ZIO
[error]   called from Main$.$$anonfun$main$1([java.lang.String)zio.ZIO
[error]   called from Main$.main([java.lang.String)scala.Unit
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   zio.stm.STM$internal$
[error]   zio.stm.STM$
[error]   Main$
```

It seems HashMap.forEach is not implemented in scala.js as for now.

**Comment by ghostdogpr** [comment]
@jdegoes was the performance gain significant when switching from MutableMap to Hashmap? Is it worth extracting it to a class with different implementations for JVM/JS?

**Comment by jdegoes** [comment]
@ghostdogpr We don't need to use `forEach` on that; we can stick with the Java version but just do it the old fashioned way with the `entrySet`. It's slightly slower but will be less overhead than the Scala version.

/cc @heraklos

**Comment by heraklos** [comment]
@ghostdogpr 
@jdegoes 
Thank you for clarifying the way it should be!

I just submitted PR, so please give me your review when you are available :)
https://github.com/zio/zio/pull/1338

**Comment by ghostdogpr** [comment]
Closed by #1338
